pipeline-steps/common.groovy

```diff
@@ -0,0 +1,112 @@
+    if [[ ! -d ".venv" ]]; then
+        virtualenv --python=/opt/rh/python27/root/usr/bin/python .venv
+    fi
+    # hack the selinux module into the venv
+    cp -r /usr/lib64/python2.6/site-packages/selinux .venv/lib64/python2.7/site-packages/
+    source .venv/bin/activate
+
+    # These pip commands cannot be combined into one.
+    pip install -U six packaging appdirs
+    pip install -U setuptools pip
+    pip install -U ansible pyrax
+  """
+}
+
+/* Run ansible-playbooks within a venev
+ * Sadly the standard ansibleplaybook step doesn't allow specifying a custom
+ * ansible path. It does allow selection of an ansible tool, but those are
+ * statically configured in global jenkins config.
+ *
+ * Args:
+ *  playbooks: list of playbook filenames
+ *  vars: dict of vars to be passed to ansible as overrides
+ *  args: list of string args to pass to ansible-playbook
+ *  venv: path to venv to activate before runing ansible-playbook.
+ */
+def venvPlaybook(Map args){
+  withEnv(['ANSIBLE_FORCE_COLOR=true',
+           'ANSIBLE_HOST_KEY_CHECKING=False']){
+    ansiColor('xterm'){
+      if (!('vars' in args)){
+        args.vars=[:]
+      }
+      for (i=0; i<args.playbooks.size(); i++){
+        playbook = args.playbooks[i]
+        vars_file="vars.${playbook}"
+        write_json(file: vars_file, obj: args.vars)
+        sh """
+          which scl && source /opt/rh/python27/enable
+          . ${args.venv}/bin/activate
+          ansible-playbook -vvv ${args.args.join(' ')} -e@${vars_file} ${playbook}
+        """
+      } //for
+    } //color
+  } //withenv
+} //venvplaybook
+
+/*
+ * JsonSluperClassic and JsonOutput are not serializable, so they
+ * can only be used in @NonCPS methods. However readFile and writeFile
+ * cannotbe used in NonCPS methods, so reading and writing json
+ * requires one function to handle the io, and another to do the.
+ * conversion.
+ *
+ * JsonSluperClassic returns a serializable object, but JsonSlurper
+ * does not. This makes Classic preferable for pipeline use.
+ */
+@NonCPS
+def _parse_json_string(Map args){
+  return (new JsonSlurperClassic()).parseText(args.json_text)
+}
+
+@NonCPS
+def _write_json_string(Map args){
+    return (new JsonOutput()).toJson(args.obj)
+}
+
+/* Read Json file and return object
+ * Args:
+ *  file: String path of file to read
+ */
+def parse_json(Map args){
+    return this._parse_json_string(
+      json_text: readFile(file: args.file)
+    )
+}
+
+/* Write object to file as JSON
+ * Args:
+ *  file: String path of file to write
+ *  obj: Object to translate into JSON
+ */
+def write_json(Map args){
+  writeFile(
+    file: args.file,
+    text: this._write_json_string(obj: args.obj)
+  )
+}
+
+/* Run a stage if the stage name is contained in an env var
+ * Args:
+ *   - stage_name: String name of this stage
+ *   - stage: Closure to execute
+ * Environment:
+ *    - STAGES: String list of stages that should be run
+ */
+def conditionalStage(Map args){
```

Comment:
So there might be a drawback with doing it this way because the Jenkins stage view can't handle dynamic stages. Basically, if the structure of the stages change, it'll reset the stage view and we lose the nice stage view of the older runs. For example, if you previously had `Allocate | Prepare | Deploy | Test | Cleanup` currently in the stage view and you do a new build with just `Prepare | Deploy | Test | Cleanup`, the stage view will no longer show the old build and only have the latest. It's okay if it's just missing the last stages like `Allocate | Prepare | Deploy | Test`, though.

It's not a huge deal since it's just a visual aid, but it's maybe something to consider.



https://wiki.jenkins-ci.org/display/JENKINS/Pipeline+Stage+View+Plugin

> Dynamic stages: in general, if you want to visualize dynamically changing stages, make it conditional to execute the stage contents, not conditional to include the stage** Stage view can handle a limited subset of cases where you're appending additional stages, but the cell-based view is not matched to changing stage structures